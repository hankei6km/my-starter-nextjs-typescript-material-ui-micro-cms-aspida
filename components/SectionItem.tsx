import React, { ElementType } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import Typography, { TypographyProps } from '@material-ui/core/Typography';
import ArticleList from './ArticleList';
import { Section as SectionType } from '../types/pageTypes';
import { pruneClasses } from '../utils/classes';
import SectionContext, { sectionContextDefault } from './SectionContext';

const useStyles = makeStyles(() => ({
  'SectionItem-root': {
    width: '100%'
  },
  'SectionItem-title': {
    width: '100%'
  },
  'SectionItem-contentBody': {
    width: '100%'
  },
  'SectionItem-contentArticles': {
    width: '100%'
  }
}));
const classNames = [
  'SectionItem-root',
  'SectionItem-title',
  'SectionItem-contentBody',
  'SectionItem-contentArticles'
];

export type SectionItemComponent = {
  sectionTitleComponent: ElementType<any>;
};

export type SectionItemVariant = {
  sectionTitleVariant: TypographyProps['variant'];
};

type Props = {
  data: SectionType;
  classes?: { [key: string]: string };
};

const SectionItem = ({ data, classes: inClasses }: Props) => {
  const classes = useStyles({ classes: pruneClasses(inClasses, classNames) });
  const { component, variant } = sectionContextDefault;
  return (
    <SectionContext.Provider value={sectionContextDefault}>
      <Box component="section" className={classes['SectionItem-root']}>
        {data.title && (
          <Typography
            variant={variant.sectionTitleVariant}
            component={component.sectionTitleComponent}
            className={classes['SectionItem-title']}
          >
            {data.title}
          </Typography>
        )}
        {data.content.map((content, i) => (
          <Box key={i}>
            {content.kind === 'html' && (
              <Box
                className={classes['SectionItem-contentBody']}
                dangerouslySetInnerHTML={{ __html: content.contentHtml }}
              ></Box>
            )}
            {content.kind === 'posts' && (
              <Box className={classes['SectionItem-contentArticles']}>
                <ArticleList
                  items={content.contents}
                  detail={content.detail}
                  classes={{ ...inClasses }}
                />
              </Box>
            )}
          </Box>
        ))}
      </Box>
    </SectionContext.Provider>
  );
};
export default SectionItem;
