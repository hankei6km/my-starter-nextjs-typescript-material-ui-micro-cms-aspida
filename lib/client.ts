import aspida from '@aspida/fetch';
import api from '../types/client/$api';
import mock from '../types/client/$mock';

export const fetchConfig = (() => {
  const getApiKey = process.env.GET_API_KEY || '';
  if (getApiKey === '' && process.env.NODE_ENV !== 'test') {
    console.error('$GET_API_KEY is not defined.');
  }
  return {
    headers: { 'X-API-KEY': getApiKey }
  };
})();

// aspida の client.xxx の xxx 部分。
// 今回は api.vi[apiName]
// 手動で定義する必要があるので注意:
export const ApiNameArticleValues = ['posts'] as const;
export type ApiNameArticle = typeof ApiNameArticleValues[number];

const clientV1 = (process.env.USE_MOCK_CLIENT_FORCE === 'true' ||
process.env.USE_MOCK_CLIENT === 'true'
  ? mock(aspida(fetch))
  : api(aspida(fetch))
).api.v1;

export default clientV1;
