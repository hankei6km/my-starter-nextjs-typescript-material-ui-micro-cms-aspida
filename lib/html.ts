import unified, { Processor, FrozenProcessor } from 'unified';
import rehypeParse from 'rehype-parse';
import rehypeMinifyWhitespace from 'rehype-minify-whitespace';
import rehypeStringify from 'rehype-stringify';
import rehypeSanitize from 'rehype-sanitize';
import merge from 'deepmerge';
import gh from 'hast-util-sanitize/lib/github.json';
import { Schema } from 'hast-util-sanitize';
import cheerio from 'cheerio';
import parseStyle from 'style-to-object';
// import camelcaseKeys from 'camelcase-keys';  // vedor prefix が jsx styleにならない?
import { Section, SectionContentHtmlChildren } from '../types/pageTypes';

import visit from 'unist-util-visit';
import { Element } from 'hast';
// import { Heading } from 'mdast';
import { Transformer } from 'unified';

const schema = merge(gh, {
  tagNames: ['picture', 'source'],
  attributes: {
    source: ['srcSet', 'sizes'],
    img: ['srcSet', 'sizes', 'className']
  }
});

export function adjustHeading(
  { top }: { top: number } = { top: 4 }
): Transformer {
  // hコンテント HTML と Markdown で本文に入力される見出しは h2 と h3 .
  // h1 は保険で最上位にまるめる、.
  return function transformer(tree): void {
    // 今回は最上位が h3 か h4 に固定
    function visitorH3Top(node: Element, _index: number): void {
      switch (node.tagName) {
        case 'h1':
          node.tagName = 'h3';
          break;
        case 'h2':
          node.tagName = 'h3';
          break;
        case 'h3':
          node.tagName = 'h4';
          break;
      }
    }
    function visitorH4Top(node: Element, _index: number): void {
      switch (node.tagName) {
        case 'h1':
          node.tagName = 'h4';
          break;
        case 'h2':
          node.tagName = 'h4';
          break;
        case 'h3':
          node.tagName = 'h5';
          break;
      }
    }
    visit(tree, 'element', top === 3 ? visitorH3Top : visitorH4Top);
  };
}

export function processorHtml() {
  return unified().use(rehypeParse, { fragment: true });
}

function normalizeProcessor(processor: Processor): FrozenProcessor {
  return processor
    .use(rehypeMinifyWhitespace)
    .use(rehypeSanitize, (schema as unknown) as Schema)
    .use(rehypeStringify)
    .freeze();
}

export function normalizedHtml(processor: Processor, html: string): string {
  let ret = '';
  normalizeProcessor(processor).process(html, (err, file) => {
    if (err) {
      console.error(err);
    }
    ret = String(file);
  });
  return ret;
}

export function styleToJsxStyle(
  s: string
): SectionContentHtmlChildren['style'] {
  const p = parseStyle(s) || {};
  const ret: SectionContentHtmlChildren['style'] = {};
  Object.entries(p).forEach((kv) => {
    const [first, ...names] = kv[0].split('-');
    const capitalizedNames = names.map(
      (r) => `${r[0].toLocaleUpperCase()}${r.slice(1)}`
    );
    const k = `${first}${capitalizedNames.join('')}`;

    ret[k] = kv[1];
  });
  return ret;
}

// とりあえず、普通に記述された markdown から変換されたときに body の直下にありそうなタグ.
// いまのところ小文字のみ.
export const SectionContentHtmlChildrenElemValues = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'div',
  'p',
  'blockquote',
  'code',
  'pre',
  'span',
  'img',
  'picture',
  'table',
  'ul',
  'ol',
  'dl',
  'hr',
  'a',
  'header',
  'footer',
  'section',
  'article',
  'aside'
] as const;
export function htmlToChildren(html: string): SectionContentHtmlChildren[] {
  // markdown の変換と sanitize には unifed を使っているので、少し迷ったが cheerio を利用.
  // (unified は typescript で利用するときに対応していないプラグインがたまにあるので)

  const ret: SectionContentHtmlChildren[] = [];
  const $ = cheerio.load(html);
  try {
    $('body')
      .contents()
      .each((_idx, elm) => {
        // https://stackoverflow.com/questions/31949521/scraping-text-with-cheerio
        // if (elm.nodeType === Node.ELEMENT_NODE) {
        if (elm.nodeType === 1) {
          if (
            SectionContentHtmlChildrenElemValues.some((v) => v === elm.tagName)
          ) {
            const $elm = $(elm);
            const attribs = elm.attribs ? { ...elm.attribs } : {};
            let style: SectionContentHtmlChildren['style'] = {};
            if (elm.attribs.style) {
              style = styleToJsxStyle(elm.attribs.style);
              delete attribs.style;
            }
            if (elm.attribs.class) {
              attribs.className = elm.attribs.class;
              delete attribs.class;
            }
            ret.push({
              tagName: elm.tagName,
              style: style,
              attribs: attribs,
              html: $elm.html() || ''
            });
          } else {
            throw new Error(
              `support only ${SectionContentHtmlChildrenElemValues.join(', ')}`
            );
          }
        } else {
          throw new Error('support only nodeType=Node.ELEMENT_NODE');
        }
      });
  } catch (_e) {
    ret.splice(0, ret.length, {
      tagName: 'div',
      style: {},
      attribs: {},
      html: html
    });
  }
  if (ret.length === 0) {
    ret.push({
      tagName: 'div',
      style: {},
      attribs: {},
      html: html
    });
  }
  return ret;
}

export type IndexedHtml = {
  index: {
    range: [
      // start and end pos.
      number,
      number
    ];
    // section  section contentHtml を特定する idx.
    sectionIdx: number;
    contentIdx: number;
    childIdx: number;
  }[];
  html: string;
};

export function getIndexedHtml(sections: Section[]): IndexedHtml {
  const ret: IndexedHtml = {
    index: [],
    html: ''
  };
  sections.forEach((section, sectionIdx) => {
    section.content.forEach((content, contentIdx) => {
      if (content.kind === 'html') {
        // filter ではじくと index がずれる
        // const contentHtmlWithoutCodeBlock = content.contentHtml.filter(
        //   ({ tagName, html }) =>
        //     tagName !== 'pre' && html.slice(0, 4) !== 'code'
        // );
        content.contentHtml.forEach(({ tagName, html }, childIdx) => {
          if (
            tagName !== 'img' &&
            !(tagName === 'pre' && html.slice(0, 5) === '<code')
          ) {
            const start = ret.html.length;
            ret.html = `${ret.html}<${tagName}>${html}</${tagName}>`;
            const index: IndexedHtml['index'][0] = {
              range: [start, ret.html.length - 1],
              sectionIdx: sectionIdx,
              contentIdx: contentIdx,
              childIdx: childIdx
            };
            ret.index.push(index);
          }
        });
      }
    });
  });
  return ret;
}
