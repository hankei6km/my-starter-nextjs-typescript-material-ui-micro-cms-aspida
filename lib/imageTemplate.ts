import { ContentImage } from './image';
import { SrcAttr, PictureNode, ImgNode } from './intermediate';

export type ImageMatcher = (content: ContentImage) => boolean;

// サムネイル画像をクリックしたときに開く画像のサイズ再現に使う.
// (ImagteTemplate はタグを作成用なので、href のような場所には向かない)
export type ImageTransform = {
  matcher: ImageMatcher;
  paramsStr: string;
};

export type ImageTemplate = {
  matcher: ImageMatcher;
  intermediate: PictureNode | ImgNode;
  asThumb: boolean;
  largeImage?: ImageTransform;
  overrideParams?: SrcAttr['url']['params'];
  templateSrc?: string;
};

export const imageMatcherLandscape = (
  min: number,
  max: number = -1
): ImageMatcher => {
  return (content: ContentImage): boolean => {
    if (content.image.width >= content.image.height) {
      return (
        min <= content.image.width && (max < 0 || content.image.width <= max)
      );
    }
    return false;
  };
};

export const imageMatcherPortrait = (
  min: number,
  max: number = -1
): ImageMatcher => {
  return (content: ContentImage): boolean => {
    if (content.image.height > content.image.width) {
      return (
        min <= content.image.height && (max < 0 || content.image.height <= max)
      );
    }
    return false;
  };
};

export const imageMatcherOr = (
  m1: ImageMatcher,
  m2: ImageMatcher
): ImageMatcher => {
  return (content: ContentImage): boolean => {
    return m1(content) || m2(content);
  };
};

export const imageMatcherTrue = (): ImageMatcher => {
  return (_content: ContentImage): boolean => {
    return true;
  };
};

export function imageTransformedUrl(
  content: ContentImage,
  imageOpenedSize: ImageTransform
): string {
  if (imageOpenedSize.matcher(content)) {
    return `${content.image.url}?${imageOpenedSize.paramsStr}`;
  }
  return content.image.url;
}
