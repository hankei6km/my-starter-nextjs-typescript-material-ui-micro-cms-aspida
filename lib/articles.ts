import { ParsedUrlQuery } from 'querystring';
import { GetStaticPropsContext } from 'next';
import client, { fetchConfig, ApiNameArticle } from './client';
import { ArticlePost, blankArticlePost } from '../types/articleTypes';
import {
  ArticleContent,
  blankArticleContent
} from '../types/client/contentTypes';
import { markdownToHtml } from './markdown';

export async function getSortedArticleList(apiName: ApiNameArticle) {
  try {
    const res = await client[apiName].get({
      query: {
        fields: 'id,createdAt,updatedAt,publishedAt,revisedAt,title'
      },
      config: fetchConfig
    });
    return res.body.contents;
  } catch (err) {
    // res.status = 404 などでも throw される(試した限りでは)
    // res.status を知る方法は?
    console.error(`getSortedArticleList error: ${apiName}: ${err.name}`);
  }
  return [];
}

export async function getAllArticleIds(apiName: ApiNameArticle) {
  try {
    const res = await client[apiName].get({
      query: {
        fields: 'id'
      },
      config: fetchConfig
    });
    return res.body.contents.map(({ id }) => ({ params: { id } }));
  } catch (err) {
    console.error(`getAllArticleIds error: ${apiName}: ${err.name}`);
  }
  return [];
}

export async function getArticleData(
  apiName: ApiNameArticle,
  {
    params = { id: '' },
    preview = false,
    previewData = {}
  }: GetStaticPropsContext<ParsedUrlQuery>
): Promise<ArticleContent> {
  try {
    const res = await client[apiName]
      ._id(!preview ? params.id : previewData.slug) // 似たような3項式がバラけていてすっきりしない
      .$get({
        query: {
          draftKey: !preview ? '' : previewData.draftKey
        },
        config: fetchConfig
      });
    return res;
  } catch (err) {
    console.error(`getArticleData error: ${apiName}: ${err.name}`);
  }
  return blankArticleContent();
}

export async function getArticlePostData(
  apiName: ApiNameArticle,
  {
    params = { id: '' },
    preview = false,
    previewData = {}
  }: GetStaticPropsContext<ParsedUrlQuery>
): Promise<ArticlePost> {
  try {
    const res = await getArticleData(apiName, {
      params,
      preview,
      previewData
    });
    const data = {
      id: res.id,
      updated: res.revisedAt || res.updatedAt,
      title: res.title,
      content: res.content.map((content) => {
        if (content.fieldId === 'contentHtml') {
          return { kind: 'html' as 'html', html: content.html };
        } else if (content.fieldId === 'contentMarkdown') {
          return {
            kind: 'html' as 'html',
            html: markdownToHtml(content.markdown)
          };
        }
        return { kind: 'html' as 'html', html: '' };
      }),
      mainImage: res.mainImage || ''
    };
    return data;
  } catch (err) {
    console.error(`getArticlePostData error: ${apiName}: ${err.name}`);
  }
  return blankArticlePost();
}
