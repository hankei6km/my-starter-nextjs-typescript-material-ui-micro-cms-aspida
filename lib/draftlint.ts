import cheerio from 'cheerio';
import { TextlintKernel } from '@textlint/kernel';
import { TextlintKernelOptions } from '@textlint/kernel/lib/textlint-kernel-interface';
import { Section } from '../types/pageTypes';
import { getIndexedHtml } from './html';
import { getTextlintKernelOptions } from '../utils/textlint';
import traverse from 'traverse';
const parseHtml = require('textlint-plugin-html/lib/html-to-ast').parse;

type InsInfo = {
  message: string;
  ruleId: string;
  severity: number;
  index: number;
  insIndex: number;
  range: [number, number];
};
type InsInfos = InsInfo[];

export function getInsInfos(html: string, messages: any[]): InsInfos {
  const ret: InsInfos = [];
  const nodes = traverse(parseHtml(html)).nodes();
  messages.forEach((m) => {
    let hit = -1;
    const index = m.index;
    nodes.forEach((node, ii) => {
      if (
        node.type &&
        node.range &&
        node.range[0] <= index &&
        index <= node.range[1]
      ) {
        hit = ii;
      }
    });
    if (hit >= 0) {
      const node = nodes[hit];
      const end = m.index - node.range[0];
      // console.log(end, node.value.slice(0, end), node.raw.slice(0, end));
      ret.push({
        message: m.message,
        ruleId: m.ruleId,
        severity: m.severity,
        index: m.index,
        // 末尾 ';' だと欠ける? よって先頭を使う
        // TODO: 末尾の扱いチェック(どこで欠ける?).
        insIndex:
          node.value &&
          node.raw &&
          node.value.slice(0, end) === node.raw.slice(0, end)
            ? m.index
            : node.range[0] - 1,
        range: [node.range[0] as number, node.range[1] as number]
      });
    }
  });
  return ret.sort((a, b) => a.insIndex - b.insIndex);
}

export function insertHtmlToSections(
  html: string,
  pos: number, // section を IndexedHtml の html に変換した場合の html の位置
  sections: Section[]
): Section[] {
  //  const ret = [...sections];
  const { index } = getIndexedHtml(sections);
  const idx = index.findIndex(
    ({ range }) => range[0] <= pos && pos <= range[1]
  );
  if (idx >= 0) {
    const { range, sectionIdx, contentIdx, childIdx } = index[idx];
    const content = sections[sectionIdx].content[contentIdx];
    if (content.kind === 'html') {
      const c = content.contentHtml[childIdx];
      let posInChild = pos - range[0] - (c.tagName.length + 2); // 2 = '<>'.length
      posInChild = posInChild < 0 ? 0 : posInChild;
      c.html = `${c.html.slice(0, posInChild)}${html}${c.html.slice(
        posInChild
      )}`;
    }
  } // 見つからないときは?
  return sections;
}
type TextLintInSectionsResult = {
  sections: Section[];
  messages: { ruleId: string; message: string; id: string; severity: number }[];
  list: string;
};

export async function textLintInSections(
  sections: Section[],
  options?: TextlintKernelOptions,
  messageStyle: { [key: string]: string } = { color: 'red' },
  idPrefix: string = ''
): Promise<TextLintInSectionsResult> {
  let ret: TextLintInSectionsResult = { sections, messages: [], list: '' };
  const indexedHtml = getIndexedHtml(sections);

  // https://github.com/mobilusoss/textlint-browser-runner/tree/master/packages/textlint-bundler
  const kernel = new TextlintKernel();

  const results = [
    await kernel.lintText(
      indexedHtml.html,
      // todo: options(presets など)は SiteServerSideConfig で定義できるようにする.
      options || getTextlintKernelOptions()
    )
  ];
  if (results && results.length > 0 && results[0].messages.length > 0) {
    const insInfos = getInsInfos(indexedHtml.html, results[0].messages);
    let slider = 0;
    const $wrapper = cheerio.load('<span/>')('span');
    Object.entries(messageStyle).forEach(([k, v]) => {
      $wrapper.css(k, v);
    });
    insInfos.forEach((m, i) => {
      const id = `${idPrefix}:textLintMessage:${i}`;
      $wrapper.attr('id', id);
      const html = $wrapper.html(m.message).parent().html();
      if (m.message && html) {
        const index = m.insIndex + 1;
        ret.sections = insertHtmlToSections(html, index + slider, ret.sections);
        slider = slider + html.length;
        ret.messages.push({
          ruleId: m.ruleId,
          message: m.message,
          id,
          severity: m.severity
        });
      }
    });
    const $dl = cheerio.load('<dl/>')('dl');
    ret.messages.forEach((m) => {
      const $d = cheerio.load('<dt></dt><dd><a/></dd>');
      // https://github.com/textlint/textlint/blob/master/packages/%40textlint/kernel/src/shared/rule-severity.ts
      // https://github.com/textlint/textlint/blob/master/packages/%40textlint/kernel/src/context/TextlintRuleSeverityLevelKeys.ts
      $d('dt').text(
        m.severity === 0 ? 'info' : m.severity === 1 ? 'warning' : 'error'
      );
      const $a = $d('a');
      $a.attr('href', `#${m.id}`);
      $a.text(`${m.message}(${m.ruleId})`);
      $dl.append($d('body').children());
    });
    ret.list = $dl.parent().html() || '';
  }
  return ret;
}
