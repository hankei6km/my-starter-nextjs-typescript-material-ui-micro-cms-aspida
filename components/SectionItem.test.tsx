import React from 'react';
import { render, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { RouterContext } from 'next/dist/next-server/lib/router-context';
import {
  mockRouter,
  mockImage,
  mockDataSectionArticleList
} from '../test/testUtils';

import SectionItem from './SectionItem';

// https://stackoverflow.com/questions/40449434/mocking-globals-in-jest
const orgImage = global.Image;
afterEach(() => {
  global.Image = orgImage;
});

describe('SectionItem', () => {
  it('should renders Section by List ', async () => {
    const addEventListener = jest.fn();
    global.Image = mockImage(jest.fn(), addEventListener);

    await act(async () => {
      const router = mockRouter();
      const { container } = render(
        <RouterContext.Provider value={router}>
          <SectionItem
            data={{
              title: 'posts',
              content: [
                {
                  kind: 'posts',
                  contents: mockDataSectionArticleList,
                  detail: true
                }
              ]
            }}
          />
        </RouterContext.Provider>
      );

      await new Promise<void>((resolve) => {
        setTimeout(() => {
          addEventListener.mock.calls.forEach((v) => {
            v[1]();
          });
          resolve();
        }, 10);
      });
      const titleOuter = container.querySelector(
        `[class*='ArticleDetail-root']`
      );
      expect(titleOuter).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });
});
