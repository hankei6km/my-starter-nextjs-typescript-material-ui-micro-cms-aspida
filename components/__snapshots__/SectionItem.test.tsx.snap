// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionItem should renders Section by List  1`] = `
<div>
  <section
    class="MuiBox-root MuiBox-root-13 makeStyles-SectionItem-root-1"
  >
    <h2
      class="MuiTypography-root makeStyles-SectionItem-title-2 MuiTypography-h2"
    >
      posts
    </h2>
    <div
      class="MuiBox-root MuiBox-root-14"
    >
      <div
        class="MuiBox-root MuiBox-root-15 makeStyles-SectionItem-contentArticles-4"
      >
        <div
          class="MuiBox-root MuiBox-root-19 makeStyles-ArticleList-root-16 makeStyles-ArticleList-detailList-17"
        >
          <article
            class="MuiBox-root MuiBox-root-27 makeStyles-ArticleDetail-root-20"
          >
            <header
              class="MuiBox-root MuiBox-root-28 makeStyles-ArticleDetail-header-21"
            >
              <h3
                class="MuiTypography-root makeStyles-ArticleDetail-title-24 MuiTypography-h3 MuiTypography-colorTextPrimary"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                  href="/posts/mmmmmmmmm"
                >
                  title4
                </a>
              </h3>
              <div
                class="MuiBox-root MuiBox-root-29 makeStyles-ArticleDetail-updated-25"
              >
                <div
                  class="MuiBox-root MuiBox-root-32 makeStyles-DateUpdated-root-30"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                    <path
                      d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                    />
                  </svg>
                  <div
                    class="MuiTypography-root makeStyles-DateUpdated-date-31 MuiTypography-body2"
                  >
                    <time
                      datetime="2021-01-13T05:12.157Z"
                    >
                      2021-01-13
                    </time>
                  </div>
                </div>
              </div>
            </header>
            <div
              class="MuiBox-root MuiBox-root-33 makeStyles-ArticleDetail-content-22"
            >
              <div
                class="MuiBox-root MuiBox-root-34 makeStyles-ArticleDetail-thumbImage-26"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                  href="/posts/mmmmmmmmm"
                >
                  <div
                    class="makeStyles-thumb-35"
                  >
                    <img
                      alt="thumbnail for title4"
                      height="150"
                      width="250"
                    />
                  </div>
                </a>
              </div>
            </div>
            <footer
              class="MuiBox-root MuiBox-root-36 makeStyles-ArticleDetail-footer-23"
            >
              <a
                aria-disabled="false"
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
                href="/posts/mmmmmmmmm"
                role="button"
                style="text-transform: none;"
                tabindex="0"
              >
                <span
                  class="MuiButton-label"
                >
                  read more
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </footer>
          </article>
          <article
            class="MuiBox-root MuiBox-root-37 makeStyles-ArticleDetail-root-20"
          >
            <header
              class="MuiBox-root MuiBox-root-38 makeStyles-ArticleDetail-header-21"
            >
              <h3
                class="MuiTypography-root makeStyles-ArticleDetail-title-24 MuiTypography-h3 MuiTypography-colorTextPrimary"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                  href="/posts/zzzzzzzzz"
                >
                  title3
                </a>
              </h3>
              <div
                class="MuiBox-root MuiBox-root-39 makeStyles-ArticleDetail-updated-25"
              >
                <div
                  class="MuiBox-root MuiBox-root-40 makeStyles-DateUpdated-root-30"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                    <path
                      d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                    />
                  </svg>
                  <div
                    class="MuiTypography-root makeStyles-DateUpdated-date-31 MuiTypography-body2"
                  >
                    <time
                      datetime="2020-12-27T04:04:30.107Z"
                    >
                      2020-12-27
                    </time>
                  </div>
                </div>
              </div>
            </header>
            <div
              class="MuiBox-root MuiBox-root-41 makeStyles-ArticleDetail-content-22"
            >
              <div
                class="MuiBox-root MuiBox-root-42 makeStyles-ArticleDetail-thumbImage-26"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                  href="/posts/zzzzzzzzz"
                >
                  <div
                    class="makeStyles-thumb-35"
                  >
                    <img
                      alt="thumbnail for title3"
                      height="150"
                      width="250"
                    />
                  </div>
                </a>
              </div>
            </div>
            <footer
              class="MuiBox-root MuiBox-root-43 makeStyles-ArticleDetail-footer-23"
            >
              <a
                aria-disabled="false"
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
                href="/posts/zzzzzzzzz"
                role="button"
                style="text-transform: none;"
                tabindex="0"
              >
                <span
                  class="MuiButton-label"
                >
                  read more
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </footer>
          </article>
          <article
            class="MuiBox-root MuiBox-root-44 makeStyles-ArticleDetail-root-20"
          >
            <header
              class="MuiBox-root MuiBox-root-45 makeStyles-ArticleDetail-header-21"
            >
              <h3
                class="MuiTypography-root makeStyles-ArticleDetail-title-24 MuiTypography-h3 MuiTypography-colorTextPrimary"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                  href="/posts/yyyyyy-da"
                >
                  title2
                </a>
              </h3>
              <div
                class="MuiBox-root MuiBox-root-46 makeStyles-ArticleDetail-updated-25"
              >
                <div
                  class="MuiBox-root MuiBox-root-47 makeStyles-DateUpdated-root-30"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                    <path
                      d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                    />
                  </svg>
                  <div
                    class="MuiTypography-root makeStyles-DateUpdated-date-31 MuiTypography-body2"
                  >
                    <time
                      datetime="2020-12-26T15:29:14.476Z"
                    >
                      2020-12-26
                    </time>
                  </div>
                </div>
              </div>
            </header>
            <div
              class="MuiBox-root MuiBox-root-48 makeStyles-ArticleDetail-content-22"
            >
              <div
                class="MuiBox-root MuiBox-root-49 makeStyles-ArticleDetail-thumbImage-26"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                  href="/posts/yyyyyy-da"
                >
                  <div
                    class="makeStyles-thumb-35"
                  >
                    <img
                      alt="thumbnail for title2"
                      height="150"
                      width="250"
                    />
                  </div>
                </a>
              </div>
            </div>
            <footer
              class="MuiBox-root MuiBox-root-50 makeStyles-ArticleDetail-footer-23"
            >
              <a
                aria-disabled="false"
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
                href="/posts/yyyyyy-da"
                role="button"
                style="text-transform: none;"
                tabindex="0"
              >
                <span
                  class="MuiButton-label"
                >
                  read more
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </footer>
          </article>
          <article
            class="MuiBox-root MuiBox-root-51 makeStyles-ArticleDetail-root-20"
          >
            <header
              class="MuiBox-root MuiBox-root-52 makeStyles-ArticleDetail-header-21"
            >
              <h3
                class="MuiTypography-root makeStyles-ArticleDetail-title-24 MuiTypography-h3 MuiTypography-colorTextPrimary"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                  href="/posts/xxxxxxxxx"
                >
                  title1
                </a>
              </h3>
              <div
                class="MuiBox-root MuiBox-root-53 makeStyles-ArticleDetail-updated-25"
              >
                <div
                  class="MuiBox-root MuiBox-root-54 makeStyles-DateUpdated-root-30"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                    <path
                      d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                    />
                  </svg>
                  <div
                    class="MuiTypography-root makeStyles-DateUpdated-date-31 MuiTypography-body2"
                  >
                    <time
                      datetime="2020-12-26T12:27:22.533Z"
                    >
                      2020-12-26
                    </time>
                  </div>
                </div>
              </div>
            </header>
            <div
              class="MuiBox-root MuiBox-root-55 makeStyles-ArticleDetail-content-22"
            >
              <div
                class="MuiBox-root MuiBox-root-56 makeStyles-ArticleDetail-thumbImage-26"
              >
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
                  href="/posts/xxxxxxxxx"
                >
                  <div
                    class="makeStyles-thumb-35"
                  >
                    <img
                      alt="thumbnail for title1"
                      height="150"
                      width="250"
                    />
                  </div>
                </a>
              </div>
            </div>
            <footer
              class="MuiBox-root MuiBox-root-57 makeStyles-ArticleDetail-footer-23"
            >
              <a
                aria-disabled="false"
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-outlinedPrimary MuiButton-textSizeSmall MuiButton-sizeSmall MuiTypography-colorPrimary"
                href="/posts/xxxxxxxxx"
                role="button"
                style="text-transform: none;"
                tabindex="0"
              >
                <span
                  class="MuiButton-label"
                >
                  read more
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </footer>
          </article>
        </div>
      </div>
    </div>
  </section>
</div>
`;
